examples/developer_guide/4_rabbitmq_cpp_Stage/_lib/rabbitmq_source.cpp

```diff
@@ -0,0 +1,143 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "rabbitmq_source.hpp"
+
+#include <boost/fiber/operations.hpp>  // for this_fiber::sleep_for
+#include <cudf/io/json.hpp>
+#include <cudf/table/table.hpp>
+#include <glog/logging.h>
+#include <pybind11/chrono.h>  // for timedelta->chrono conversions
+#include <pybind11/pybind11.h>
+
+#include <exception>
+#include <sstream>
+#include <vector>
+
+namespace morpheus_rabbit {
+
+RabbitMQSourceStage::RabbitMQSourceStage(const std::string &host,
+                                         const std::string &exchange,
+                                         const std::string &exchange_type,
+                                         const std::string &queue_name,
+                                         std::chrono::milliseconds poll_interval) :
+  PythonSource(build()),
+  m_channel{AmqpClient::Channel::Create(host)},
+  m_poll_interval{poll_interval}
+{
+    m_channel->DeclareExchange(exchange, exchange_type);
+    m_queue_name = m_channel->DeclareQueue(queue_name);
+    m_channel->BindQueue(m_queue_name, exchange);
+}
+
+RabbitMQSourceStage::subscriber_fn_t RabbitMQSourceStage::build()
+{
+    return [this](rxcpp::subscriber<source_type_t> subscriber) -> void {
+        try
+        {
+            this->source_generator(subscriber);
+        } catch (const std::exception &e)
+        {
+            LOG(ERROR) << "Encountered error while polling RabbitMQ: " << e.what() << std::endl;
+            subscriber.on_error(std::make_exception_ptr(e));
+            close();
+            return;
+        }
+
+        close();
+        subscriber.on_completed();
+    };
+}
+
+void RabbitMQSourceStage::source_generator(rxcpp::subscriber<RabbitMQSourceStage::source_type_t> subscriber)
+{
+    const std::string consumer_tag = m_channel->BasicConsume(m_queue_name, "", true, false);
+    while (subscriber.is_subscribed())
+    {
+        AmqpClient::Envelope::ptr_t envelope;
+        if (m_channel->BasicConsumeMessage(consumer_tag, envelope, 0))
+        {
+            try
+            {
+                auto table   = from_json(envelope->Message()->Body());
+                auto message = MessageMeta::create_from_cpp(std::move(table), 0);
+                subscriber.on_next(std::move(message));
+            } catch (const std::exception &e)
+            {
+                LOG(ERROR) << "Error occurred converting RabbitMQ message to Dataframe: " << e.what();
+            }
+            m_channel->BasicAck(envelope);
```

Comment:
This is a minor thing but you might want to move this line above the `on_next` call. I'm not sure exactly what this does, but if it's meant to keep the connection alive in any way, then you want that to happen before the on_next which can yield.

Reply:
It's effectively the same as a commit in Kafka, my thinking was we would only ack messages that have been successfully emitted into the pipeline. 

Also the Kafka source does the same and commits after calling `on_next`.

Reply:
Ok, makes sense. Its fine as is.